### PR TITLE
Encapsulation of functions

### PR DESCRIPTION
**Pull request by thegreatgunbantoad**
tempScaleLabel updated to use parameters rather than refer to external variables
scaleTemperature updated to use parameters rather than refer to external variables

Note change of name of scaleTemperature to scaleTemperatureFromC

### CODE REVIEW

**Comment by thegreatgunbantoad** [comment]
Again not a functional update.  Not sure why linting is failing though.

**Comment by pppedrillo** [comment]
Why?

**Comment by thegreatgunbantoad** [comment]
> Why?

It makes code easier to reuse and understand.  Would also allow you to to pull the functions out of the main file and put them in a library.  A task it a bit different, like an initialisation task, but functions should generally be properly parameterised.

**Comment by pppedrillo** [comment]
> It makes code easier to reuse and understand. Would also allow you to to pull the functions out of the main file and put them in a library. A task it a bit different, like an initialisation task, but functions should generally be properly parameterised.

"Parameterization" != "to duplicate all the used global/static variables using arguments"

Why only those two functions then? Follow your logic, you'll need to refactor most of the rest of the functions which also use global variables (voltage, tilt etc etc) in the same manner, no?
And what if function use 125 global vars? Should it have 125 arguments?

And generally copying arguments on every function call causes code run slower+ extra mem use on stack - you are passing by value , not using a references.

File: pio/src/iSpindel.cpp [hunk 1]
```diff
@@ -75,25 +75,25 @@ uint32_t DSreqTime = 0;
 int16_t ax, ay, az;
 float Volt, Temperatur, Tilt, Gravity;
 
-float scaleTemperature(float t)
+float scaleTemperatureFromC(float t, uint8_t tempscale)
 {
-  if (myData.my_tempscale == TEMP_CELSIUS)
+  if (tempscale == TEMP_CELSIUS)
```
Comment (pppedrillo) [inline pio/src/iSpindel.cpp:80]: Why do not use switch here?

File: pio/src/iSpindel.cpp [hunk 2]
```diff
@@ -75,25 +75,25 @@ uint32_t DSreqTime = 0;
...
     return t + 273.15f;
   else
     return t; // Invalid value for myData.my_tempscale => default to celsius
 }
 
-String tempScaleLabel(void)
+String tempScaleLabel(uint8_t tempscale)
 {
-  if (myData.my_tempscale == TEMP_CELSIUS)
+  if (tempscale == TEMP_CELSIUS)
```
Comment (pppedrillo) [inline pio/src/iSpindel.cpp:92]: Same as above - why not switch?

**Review by pppedrillo — CHANGES_REQUESTED** [review]

File: pio/src/iSpindel.cpp (hunk 1 above)
Comment (thegreatgunbantoad) [reply pio/src/iSpindel.cpp:80]: Could do, I wasn't intending to make great swathes of changes all at once.

**Comment by thegreatgunbantoad** [comment]
> > It makes code easier to reuse and understand. Would also allow you to to pull the functions out of the main file and put them in a library. A task it a bit different, like an initialisation task, but functions should generally be properly parameterised.
> 
> "Parameterization" != "to duplicate all the used global/static variables using arguments"
> And those functions still not in the library or taken out of main file.
> 
> Generally why only those two functions were picked? Follow your logic, you'll need to refactor most of the rest of the functions which also use global variables (voltage, tilt etc etc) in the same manner, no?
> And what if function use 125 global vars? Should it have 125 arguments?
> 
> I'd also advise to pass args by reference whenever it makes sense (like here, myData.my_tempscale) and master a switch statement (instead of the if-else-if-else... spaghetti as soon as it is refactoring)
> 
> Renaming a function to "FromC" looks good.

Only picked those two as it was just a start at getting the functions parametrised.  If a function uses lots of global variables I think it might need re-writing.
Note the function that writes to the JSON could now just be passed the myData struct.  I was really hoping C++ had an iterator like Python that would have made writing to and reading from the JSON slicker.

Yes the plan was to refactor as many functions as is practical.  Globals are generally best avoided if reasonable, they can make code very hard to follow.

I sort of prefer pass by value as you know the variable can't change while the function is operating and the function can't alter the inputs.  But if you wanted several outputs then pass by ref is handy.  Is there a processor advantage to ref over val?

Agree with using a switch.

**Comment by pppedrillo** [comment]
> I was really hoping C++ had an iterator like Python that would have made writing to and reading from the JSON slicker.
> 

There is nothing to do with C++ language, but rather the concrete implementation of JSON (de)/serializer which is not a part of C++ standards.

> Yes the plan was to refactor as many functions as is practical. 

Thats my main concern. 
It can be a pull request then all the changes are done, not while work in progress. 

>Globals are generally best avoided if reasonable, they can make code very hard to follow.

There are some more serious problems with globals and statics, not the only code readability.

> I sort of prefer pass by value as you know the variable can't change while the function is operating and the function can't alter the inputs. But if you wanted several outputs then pass by ref is handy. Is there a processor advantage to ref over val?
> 

Passing by val not always guarantees what some data can't be changed.
If you pass an object by value and this object has a pointer to data, you still can change a data by this pointer.

Omitting an extra copy is one of advantages. Not a huge deal with one shot functions and PODs, but may cause performance issues if function gets called frequently and you pass a huge object(s) to it.

**Comment by PjotrekSE** [comment]
I make a struct of my globals, and pass the struct by reference if I need
to change a global, and by value if not.

**Review by universam1 — APPROVED** [review]
Thanks 👍🏻

**Comment by universam1** [comment]
@thegreatgunbantoad there are merge conflicts, mind rebasing please?

**Comment by thegreatgunbantoad** [comment]
@universam1 I'll have look later, not great at the git thing yet, so might need a hand.

**Comment by pppedrillo** [comment]
> 
> 
> @universam1 I'll have look later, not great at the git thing yet, so might need a hand.

I rebased those changes, #543 @universam1

**Comment by thegreatgunbantoad** [comment]
@pppedrillo does that mean I can just close this request then?  Sorry full of a cough and such so my brain is struggling.

**Comment by pppedrillo** [comment]
> 
> 
> @pppedrillo does that mean I can just close this request then? Sorry full of a cough and such so my brain is struggling.

@thegreatgunbantoad  Yes, if you are ok with changes from #543

**Comment by thegreatgunbantoad** [comment]
@pppedrillo frazzled brain thinks looks good.
